### ISSUE TITLE
Brim (Brim-Object gap?) doesn't account for Elephants-foot compensation...this creates a gap between brim and object

### ISSUE THREAD

**Issue by PhilBaz** [OrcaSlicer/OrcaSlicer#3720]
### OrcaSlicer Version

1.9

### OS version

win 10

### Additional system information

good

### Printer

Voron 2.4

### How to reproduce

 Add elephants-foot compensation and brim and slice.... If Brim-Object gap is zero there should be no gap. but there is! This gap is caused be Elephant foot compensation, and can be closed by reducing this to zero. 

![Screenshot 2024-01-17 103318](https://github.com/SoftFever/OrcaSlicer/assets/21558241/2ca91168-cfea-4228-9cb6-2aec3b1f2df1)

### Actual results

There is a gap

### Expected results

No gap, brim should touch, or be controlled entirely by brim-object gap

### Project file & Debug log uploads

[brim issue.zip](https://github.com/SoftFever/OrcaSlicer/files/13966621/brim.issue.zip)

### Checklist of files to include

- [x] Log file
- [X] Project file

**Comment by PhilBaz** [comment]
Just bumping this as its a proper bug and should be pretty easy to fix....I think?

**Comment by PhilBaz** [comment]
Bumping as this is a pretty major bug.

**Comment by PhilBaz** [comment]
Yes, just to clarify here are some images that show the issue. The brim should follow the (elephants foot compensation/ scaling ) base layer.
![a](https://github.com/SoftFever/OrcaSlicer/assets/21558241/0756940f-1b1e-408c-a2a0-3051afc7a1d6)
![b](https://github.com/SoftFever/OrcaSlicer/assets/21558241/ba23fcc3-05a7-474f-b005-e9bf3c37edb4)

**Comment by PhilBaz** [comment]
@cbaugher I have to think this is one of the largest actual bugs in Orca. This is primary functionality breaking.

**Comment by RoboCrafty** [comment]
![image](https://github.com/SoftFever/OrcaSlicer/assets/82392157/9008df70-2e75-410b-b5eb-7d23f805db0a)
I still have the gap even with elephant foot compensation set to 0. Brim-Object distance is 0 too. :/

**Comment by PhilBaz** [comment]
@RoboCrafty Then you must have a value in the brim offset. That's not a bug. Just the elephant foot compensation.

**Comment by RoboCrafty** [comment]
@PhilBaz I don't get it? I thought to remove the gap between the brim and the object the elephant foot compensation had to be set to 0 because of a bug? And here I have the gap even with it set to 0

**Comment by PhilBaz** [comment]
@RoboCrafty 
![Screenshot 2024-05-27 094904](https://github.com/SoftFever/OrcaSlicer/assets/21558241/7d580db5-9548-4a69-b6da-0d3efd73a7e3)

**Comment by RoboCrafty** [comment]
@PhilBaz Yesss I know that 😅, I mentioned I had that set to 0 and there still is a gap.

**Comment by PhilBaz** [comment]
@RoboCrafty Ohh, okay sorry. If you have that set to zero as well as elephants-foot then there should not be a gap. Maybe a bug then.

Just making sure you also didn't change any per-object settings?

**Comment by HeavilyModdedEnder3** [comment]
V2.1.0 Beta out already, and still not fixed. I have seen some argue that the elephants foot would cause it to bulge into the brim. But it's a simple feature which SuperSlicer has already implemented by allowing a negative brim offset. 

With this negative brim offset and a **large** elephants foot compensation, you can put a modifier down to make your own custom bottom layer patterns, like a carbon fiber pattern infill, and then the brim goes in and counts as the walls to make it look normal.

This way, OrcaSlicer could even have custom bottom surface patterns for infill! The next step after this would be to make something called a "rim" which always puts a wall around the perimeter, even if the rest of the top surface counts as a wall.
![image](https://github.com/SoftFever/OrcaSlicer/assets/171583557/7ab209ea-f36d-46dd-b812-36aab7dc01ee)
![image](https://github.com/SoftFever/OrcaSlicer/assets/171583557/6454de69-bcf2-4cf8-b4f7-1a11dce9e258)

See how the top of the model counts as a wall? The "rim" would force in another wall to act as a perimeter: making it look cleaner.

Even without the "rim" if you could put in a negative brim offset, without it saying "Out of Range", then the problem would be fixed. In addition, a larger elephants foot compensation could be added to make your own custom bottom surface pattern, with the addition of brims to act as walls. Finally, another feature which could be useful to allow for your own custom top surface patterns would be to make a "rim", which forces another perimeter on, even with walls existing.

**Comment by PhilBaz** [comment]
@HeavilyModdedEnder3 Too much and confuses the issue...please don't hijack this bug fix request, start your own thread. This is not what we are asking for. Please delete your post here and start your own thread.

The bug fix is simply that the brim will account for elephants foot compensation so the brim is not detached. That's all.

**Comment by neo-jayfeather** [comment]
https://wiki.bambulab.com/en/software/bambu-studio/parameter/elephant-foot
The last two screenshots seem to indicate that BBL does not plan on fixing this issue on their slicer. Just a note, but this does seem fixable by editing the .\src\libslic3r\brim.cpp and maybe the elephants foot compensation one (.\src\libslic3r\ElephantFootCompensation.cpp).

**Comment by PhilBaz** [comment]
@neo-jayfeather 

Ahhh, what a horrible decision... It should 'Just work'. Decoupling elephant's foot from brim is obviously the wrong approach.  Very much hope to see this corrected in Orca.

**Comment by brunetton** [comment]
Same here
![image](https://github.com/user-attachments/assets/212fcc51-e94e-4c7c-be6f-4b89729aebac)
![image](https://github.com/user-attachments/assets/d7a6957b-5213-4a2c-89cd-46a5bf8204c9)
(I made elephant foot compensation large to make the problem obvious)

**Comment by neo-jayfeather** [comment]
I was reading the [prusaslicer](https://help.prusa3d.com/article/elephant-foot-compensation_114487) wiki and they stated that elephant's foot compensation should squish the filament out enough so that the brim should be properly aligned-- I don't know if that applies, but that may be why no one (these slicers) have changed it.

**Comment by idmadj** [comment]
> I was reading the [prusaslicer](https://help.prusa3d.com/article/elephant-foot-compensation_114487) wiki and they stated that elephant's foot compensation should squish the filament out enough so that the brim should be properly aligned-- I don't know if that applies, but that may be why no one (these slicers) have changed it.

Sorta makes sense. If the brim doesn't connect, it's theoretically a sign that efc is set too high. 

That being said, elephant foot happens as the weight of the model squishes out the bottom (it is not meant to compensate for poor z-offset). By the time that happens the first layer has cooled down to the temperature of the bed rather than the nozzle's, so not much inter-line adhesion will take place. Also it's entirely possible that lifting will happen before the model gets a chance to elephant foot and connect with the brim. In both cases the brim will not be given a chance to do its job properly.

**Comment by PhilBaz** [comment]
Bump...this should be fixed.

**Comment by PhilBaz** [comment]
@neo-jayfeather No, this was originaly a Cura feature and there he implementation is correct...IMHO.  The brim should follow the EFC. Sometimes you use EFC to create a psudo chamfer. Also its nice to keep some EFC on your default profile so then using brim you always have to compensate and turn it off.

Also, i don't think its realistic to think of EFC as perfectly compensating so that it will always perfectly meet the brim. And with EFC its always better to use a little more rather than a little less.

In the end Cura is right here. It will be better if its fixed in Orca. Most people probably don't realize that its just not working currently.

**Comment by CCS86** [comment]
Beyond that, even with 0 EFC and 0 brim gap, there is still a gap:

![image](https://github.com/user-attachments/assets/d5b5a832-d2a4-4808-bc57-ce5c13f74790)

**Comment by PhilBaz** [comment]
Just cross posting this here.....

https://github.com/bambulab/BambuStudio/issues/4420

Please guys, this bug is silly.

I see on the bambu side its been labelled and assigned. Maybe bring your issues to the Bambu side, it may get fixed there then make it to Orca down stream.

**Comment by evilC** [comment]
PLEASE FIX!!

**Comment by hotellonely** [comment]
Bumping.

**Comment by JPK-MG** [comment]
Still a bug in version 2.3.0. 
I was hoping for a solution to this, as it renders EFC completely useless when using mouse ears, or brim. In hope of hacking the bug, I attempted to enter a negative number into Brim Gap - doesn't work, though from a coding perspective, it seems like a logical option... allow negative values in Brim Gap and you could counteract EFC's dominion over brims.

**Comment by dmach** [comment]
I just hit this issue too and realized that to fix it properly, brim might need to be printed with a slighly reduced layer height so the second layer won't stick to it.

```
+------+       <-- object with elephant foot compensation on
|      |
|   +--+       <-- this second layer overhang might stick to the brim from top
+___|_______   <-- brim
```

**Comment by serpent213** [comment]
Fix would be nice indeed!  👍🏻

**Comment by confuzled-dev** [comment]
dealing with this issue myself. pita

**Comment by PhilBaz** [comment]
If you are seeing this issue and agree that brim should just work, and not be dependent on another setting that is used for an entirely different purpose than please also go make some noise on the Bambu issue. 

https://github.com/bambulab/BambuStudio/issues/4420

**Comment by PhilBaz** [comment]
@confuzled-dev see above

**Comment by PhilBaz** [comment]
@serpent213 see above

**Comment by PhilBaz** [comment]
@hotellonely see above

**Comment by PhilBaz** [comment]
@evilC see above

**Comment by PhilBaz** [comment]
To be clear the brim object gap should ONLY be dependent on the brim object gap setting. It should not be dependent on elephants foot compensation which can be used for various reasons like adding a chamfer. Furthermore Elephants foot compensation would likely be a setting that would remain on all the time for a particular profile. So...why should the bring gap ever be dependent on it? that makes no sense! Cura got this right a decade ago!

**Comment by fyrak1s** [comment]
bump

**Comment by fyrak1s** [comment]
This bug is still an issue.

**Comment by PhilBaz** [comment]
Guys...Can we get this one fixed. Its a bad one.

**Comment by PhilBaz** [comment]
@SoftFever 

Sorry to ping you....can you have a look at this one. Thanks for your work!

**Comment by brunetton** [comment]
I totally stopped using OrcaSlicer due to this. And I'm 100% sure I'm not the only one, and a big amount of users are gone after hitting this problem

**Comment by kisslorand** [comment]
You all should look at the printed part for a gap, not in the slicer. If there's a gap that means you overcompensated for elephant foot or have a poorly calibrated printing process and/or printer.

**Comment by CCS86** [comment]
> You all should look at the printed part for a gap, not in the slicer. If there's a gap that means you overcompensated for elephant foot or have a poorly calibrated printing process and/or printer.

It sounds like you don't understand the issue at all. 

It's pretty simple:

brim-object gap: 0
elephant foot: 0
<img width="1366" height="1308" alt="Image" src="https://github.com/user-attachments/assets/c48e6dc4-f35c-4732-b599-d16b9bca96b8" />





brim-object gap: 0.15
elephant foot: 0
<img width="1366" height="1308" alt="Image" src="https://github.com/user-attachments/assets/694206f6-79f6-406b-8d4d-f074ed37eeab" />





brim-object gap: 0.15
elephant foot: 0.2
<img width="1366" height="1308" alt="Image" src="https://github.com/user-attachments/assets/ff9d12d9-54ae-45cd-a3a2-3199db615aed" />






All together now:
![Image](https://github.com/user-attachments/assets/0d684135-0cde-4c93-bb77-b4996e573adb)


This is with the latest nightly and a primitive cylinder. The third version should still have a 0.15 brim gap, but instead it has a 0.35 gap. Since you can't specify a negative brim gap (to ask for overlap), if you want to use a large elephant foot value, like 0.2, you will have a brim gap of 0.2 and can't force it to be less.

It's wild that we have gotten all kinds of advanced features like adaptive pressure advance, and small area flow compensation, before this bug has gotten any attention.

**Comment by kisslorand** [comment]
Just as I said, you're looking at the gap in the slicer, not looking at the printed part...
The elephant foot compensation value is making a gap from the model inwards, brim gap is making a gap from the model outwards.
There's no bug anywhere in the code, it does exactly what it is has been written for.
0.2mm inwards + 0.15mm outwards = 0.35mm gap -> where's the bug?

**Comment by CCS86** [comment]
It's wild that you are still missing the point here. Elephant foot compensation should not create any brim-object gap. But it currently does.

**Comment by PhilBaz** [comment]
Just to restate the central premise....

1. These feature were developed ages ago in Cura. They have always worked properly in Cura, and they currently work properly in Prusa Slicer...this is a bug.
2. Having the expected result **depend on an obscure and little understood interaction between two features in two different sections is obviously wrong and backward, from a user experience perspective**.
3. Brim gap should 'just work' regardless of your elephants foot setting. And regardless of the EF setting brim gap will require some experiment and tuning to get right.
4. Assuming the vast majority of people have perfectly calibrated EF for all cases is, of course, obviously wrong. If anything most will likely be under-sizing (micro chamfering) the first layer when tuning it by eye...which returns us to the point that brim gap will require some experiment and tuning in either case.

**Comment by kisslorand** [comment]
There's only one bug, you are using a feature out if its intended scope.
It's clear that arguing here is futile, you want to use a feature out of its scope and get frustrated it doesn't work the way you would want to.

> 4\. Assuming the vast majority of people have perfectly calibrated EF for all cases is, of course, obviously wrong

Everyone is entitled to have an opinion. Having an opinion doesn't make it true...
What's next? Assuming people calibrates their flow rate is also "obviously" wrong?

Since the code is working EXACTLY as intended, there is no bug.
A software bug is a defect, flaw, or error in a program’s code or design that causes the software to behave incorrectly, unexpectedly, or differently from its intended specification.

**Comment by CCS86** [comment]
> There's only one bug, you are using a feature out if its intended scope. It's clear that arguing here is futile, you want to use a feature out of its scope and get frustrated it doesn't work the way you would want to.


You are making this claim of "using a feature out of it's intended scope". Elaborate on that. I have no idea what you are talking about.

Let's distill this down to something even simpler for you... the brim-object gap is set to 0 here:

<img width="1203" height="1106" alt="Image" src="https://github.com/user-attachments/assets/2f5f46f1-a07d-4846-84a3-93134bcade67" />

Explain how this is feature is working correctly.

**Comment by PhilBaz** [comment]
> There's only one bug, you are using a feature out if its intended scope. It's clear that arguing here is futile, you want to use a feature out of its scope and get frustrated it doesn't work the way you would want to.
> 
> > 4. Assuming the vast majority of people have perfectly calibrated EF for all cases is, of course, obviously wrong
> 
> Everyone is entitled to have an opinion. Having an opinion doesn't make it true... What's next? Assuming people calibrates their flow rate is also "obviously" wrong?
> 
> Since the code is working EXACTLY as intended, there is no bug. A software bug is a defect, flaw, or error in a program’s code or design that causes the software to behave incorrectly, unexpectedly, or differently from its intended specification.


It was just merged to the main branch as a bug fix.

![Image](https://github.com/user-attachments/assets/65395ae2-fb71-4e7d-ab9a-2820593a79cb)

**Comment by kisslorand** [comment]
> It was just merged to the main branch as a bug fix.

I don't think so.

---

@CCS86 
I think the core misunderstanding here comes from mixing three different things that happen at different stages:
(1) first-layer physical deformation,
(2) elephant-foot compensation as a counter-deformation, and
(3) brim generation as an adhesion aid.

Elephant foot compensation is not meant to “add a chamfer” in the final part. It is a **process compensation**: the slicer deliberately undercuts or shrinks the bottom outline so that, after first-layer squish and thermal spread, the printed part **ends up to the original model geometry**. The chamfer you see in preview is not a design feature — **it’s a corrective toolpath**.

Because of that, the slicer preview must be interpreted carefully. The preview correctly shows the toolpath (yes, there is a visible gap between brim and compensated outline), but that gap is not a reliable predictor of the finished part’s base shape. In reality, the first layer tends to spread outward, often filling that space. That is exactly what EFC is compensating for.

The brim, on the other hand, is not a deformation compensation. It is an **adhesion structure**, generated **from the nominal model footprint**. Keeping it independent from EFC is intentional:
- it avoids welding the brim into the compensated edge,
- it keeps brim removal predictable,
- and it avoids over-compensating the base geometry.

So when EFC is enabled:
- the object outline is intentionally pulled inward,
- the brim remains tied to the original footprint,
- and a gap appears in preview.

That is expected behavior, not a failure of the brim-object gap logic.

“Brim-object gap = 0” means **zero gap relative to the nominal outline**, not “force the brim to follow all post-processing geometry corrections.” Forcing brim contact with compensated geometry would defeat the purpose of EFC and reintroduce the very artifacts it is meant to prevent.

If the priority is guaranteed brim contact for adhesion, then EFC should be reduced or disabled, or another adhesion method used. If the priority is dimensional accuracy at the bottom edge, then the preview gap is a natural and acceptable consequence.

In short: elephant foot compensation and brim generation solve different problems at different levels, and keeping them decoupled is the correct default behavior.

---

One more clarification: the preview correctly shows the compensated toolpath, but elephant-foot compensation is a counter-deformation. The real first layer spreads due to squish, and the undercut is meant to cancel that so the final printed part matches the original model. So a visible gap in preview does not imply a gap in the finished print.

**Comment by CCS86** [comment]
Yes, I understand that, and in a perfect world it would work that way. But, 3D printing is far from perfect. Sometimes the amount of EFC needed to correct the bottom of the model ends up leaving you with a brim that isn't attached. At least allowing a negative brim separation value would let people tune to their liking.

**Comment by kisslorand** [comment]
> Sometimes the amount of EFC needed to correct the bottom of the model ends up leaving you with a brim that isn't attached.

That cannot be true with proper settings.



> At least allowing a negative brim separation value would let people tune to their liking.

I actually proposed that already.

**Comment by brunetton** [comment]
> Yes, I understand that, and in a perfect world it would work that way. But, 3D printing is far from perfect. Sometimes the amount of EFC needed to correct the bottom of the model ends up leaving you with a brim that isn't attached. At least allowing a negative brim separation value would let people tune to their liking.

Absolutely +1

**Comment by PhilBaz** [comment]
https://github.com/OrcaSlicer/OrcaSlicer/pull/11699#issuecomment-3688215964

**Comment by PhilBaz** [comment]
For those of you who wish to see Orcas handling of brim gap brought in line with Cura Prusa etc...

https://github.com/OrcaSlicer/OrcaSlicer/pull/11699#issuecomment-3690462374

**Comment by kisslorand** [comment]
I would stop glorifying Cura in its relation between EFC and brim.

---

First layer in Cura with `Initial Layer horizontal Expansion` set to -1.0mm:

<img width="884" height="827" alt="Image" src="https://github.com/user-attachments/assets/d5042caa-b706-4bf1-b0ae-5c8a23f86e71" />

---

First layer in Cura with `Initial Layer horizontal Expansion` set to 0.0mm:
<img width="894" height="834" alt="Image" src="https://github.com/user-attachments/assets/bb35b548-0084-4428-a538-2a2000fa9892" />

---

I am not very confident that anyone wants this in Orca also.

**Comment by PhilBaz** [comment]
@kisslorand 

You are changing the wrong setting.... 

Its okay, its done. They are putting a check box.

**Comment by PhilBaz** [comment]
Done... Feature will be integrated as a check box.

https://github.com/OrcaSlicer/OrcaSlicer/pull/11699#issuecomment-3690462374

**Comment by kisslorand** [comment]
> You are changing the wrong setting....

Which ones are the "right" settings in Cura?

**Comment by PhilBaz** [comment]
@kisslorand 

What are you trying to accomplish?  The brim follows the compensation, no?

**Comment by kisslorand** [comment]
So, which ones are the "right" settings?

**Comment by PhilBaz** [comment]
@kisslorand 

Well to be relevant to our discussion you need to adjust the brim offset. But the brim is following the compensation, no?

**Comment by kisslorand** [comment]
You really cannot see anything wrong? Really?

**Comment by PhilBaz** [comment]
@kisslorand 

Like we were ever talking about geometry edge cases? That's a terrible case of semantic drift my dude.

**Comment by kisslorand** [comment]
You failed to answer 2 questions:
1. Which are the right settings?
2. You cannot see anything wrong in the screenshots?

**Comment by PhilBaz** [comment]
@kisslorand 

Oh god....dude I have kids.

Listen, you are very, very smart, and I am very, very dumb. I see it now. You showed me the truth.

**Comment by kisslorand** [comment]
I have kids too :)
...and I'm not smart :( 
(I would have kept my mouth shut if I was smart!)

**Comment by PhilBaz** [comment]
@kisslorand 

Fair.

Enjoy the new check box!

**Comment by CCS86** [comment]
A tangent discussion:

https://github.com/OrcaSlicer/OrcaSlicer/issues/11750
